src/argilla/server/search_engine/elasticsearch.py

```diff
@@ -0,0 +1,135 @@
+#  Copyright 2021-present, the Recognai S.L. team.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import dataclasses
+from typing import Any, AsyncGenerator, Dict, List, Optional
+
+from elasticsearch8 import AsyncElasticsearch, helpers
+from pydantic import conint
+
+from argilla.server.models import Dataset, Record, VectorSettings
+from argilla.server.search_engine import (
+    SearchEngine,
+    SearchResponses,
+    UserResponseStatusFilter,
+)
+from argilla.server.search_engine.commons import BaseElasticAndOpenSearchEngine, field_name_for_vector_settings
+from argilla.server.settings import settings
+
+
+def _compute_num_candidates_from_k(k: int):
+    if k < 50:
+        return 500
+    elif 50 <= k < 200:
+        return 100
+    return 2000
+
+
+@dataclasses.dataclass
+class ElasticSearchEngine(BaseElasticAndOpenSearchEngine):
+    config: Dict[str, Any]
+
+    es_number_of_shards: int
+    es_number_of_replicas: int
+
+    def __post_init__(self):
+        self.client = AsyncElasticsearch(**self.config)
+
+    def _configure_index_settings(self):
+        return {
+            "number_of_shards": self.es_number_of_shards,
+            "number_of_replicas": self.es_number_of_replicas,
+        }
+
+    def _mapping_for_vector_settings(self, vector_settings: VectorSettings) -> dict:
+        return {
+            f"vectors.{vector_settings.id}": {
+                "type": "dense_vector",
+                "dims": vector_settings.dimensions,
+                "index": True,
+                # can similarity property also be part of config @frascuchon ?
+                # relates vector search similarity metric
+                "similarity": "l2_norm",  ## default value regarding the knn best practices es documentation
```

Comment:
I think that we should let the user configure this when creating settings for a collection of vectors. In some cases, depending on the vectors used, one metric could be better than other.